### ISSUE TITLE
[Question] Can anyone reproduce ScienceQA result with the latest code?

### ISSUE THREAD

**Issue by CupidJay** [haotian-liu/LLaVA#355]
### Question

I cloned the latest code and I have tried various versions of the LLaMA-13B weights (including the original one). However, I can not reproduce the results on ScienceQA following the authors' instructions. I wonder if anyone has reproduced the results with the latest version of the code? I saw that it was successfully reproduced before, but I don't know what version it was.

Some examples of output I have here are as follows
```
    "16410": "Saint Louis is a city in Missouri.\n The answer is A is the capital of Missouri.",
    "16421": "LECTURE: Measurements are used to estimate how long something is. There are many different ways to estimate distance, but when you are trying to estimate how long something is, it is important to be accurate.\nThere are 1000 miles in 1 kilometer. So, 1000 miles is much longer than 1 mile.\nSOLUTION: The Amazon River is about 4,000 miles long.\n The answer is C.",
    "16471": "45 milliliters is a very small amount of liquid. It takes just 45 liters to fill a salt shaker.\n The answer is A.",
    "16534": "LECTURE: In the United States, the ball is in Dylan's court. This means that Dylan has some difficult work to do.\nThe idiom in this text suggests that Dylan needs to act next. This means that Dylan has to do something.\nSOLUTION: The idiom in this text suggests that Dylan has to do something. This could mean that Dylan has to act next or that he has to do some difficult work.\n The answer is A.",
    "16557": "This sentence is exclamatory. It uses an exclamatory phrase to show strong feeling or emotion.\n The answer is A.",
    "16626": "LECTURE: Inherited traits are passed down through families. Inherited traits are caused by the combination of genes from both parents.\nGenes contain information about how an organism looks and acts. Some genes contain information about how an organism should look or act.\nGenes affect how an organism looks or acts.\nGenes are passed down through families.\nChildren grow and develop into adults.\nSOLUTION: Inherited traits are passed down through families. Children grow into adults.\n The answer is A.",
    "16629": "LECTURE: In order to keep the peace, we must know what we are trying to keep. If we are trying to keep the peace, we must know what we are trying to keep. If we are trying to keep the peace, we must know what we are trying to keep.\nSOLUTION: In order to keep the peace, we must know what we are trying to keep. If we are trying to keep the peace, we must know what we are trying to keep. If we are trying to keep the peace, we must know what we are trying to keep.\n The answer is A is trying to keep the peace.",
    "16699": "This is a food Web. It is a place where organisms can find food.\n The answer is A food web is a place where organisms can find food.",
    "16744": "South Carolina is a state in the United States.\n The answer is A",
    "16810": "The euphemism is used to describe a polite way of saying that something is overweight.\n The answer is A.",
    "16878": "LECTURE: A letter is a written document that is sent to someone. It can include information about a topic or a person.\nFor example, you might write a letter to a friend, thanking him or her for a gift.\nIn the United States, the Postal Service provides a way to send letters. You can write a letter to your friend, thanking him or her for a gift.\nSOLUTION: The second sentence is correct. It thank you is the correct way to say thank you.\n The answer is A.",
```

I have been trying for two weeks, can anyone help me？

**Comment by haotian-liu** [comment]
Hi @CupidJay, I will try reproducing this again on Thursday, and update the docs if needed.

I will also release a checkpoint that contains the directly merged weights for ScienceQA.

Sorry for the confusion, and stay tuned for the update. Thanks!

**Comment by CupidJay** [comment]
Thanks for your reply. I appreciate your efforts for this great work and look forward to the update.

**Comment by CupidJay** [comment]
> Hi @CupidJay, I will try reproducing this again on Thursday, and update the docs if needed.
> 
> I will also release a checkpoint that contains the directly merged weights for ScienceQA.
> 
> Sorry for the confusion, and stay tuned for the update. Thanks!

Hi, I just cloned [this previous commit ](https://github.com/haotian-liu/LLaVA/tree/08759473173cc0a9fdc852f06a04c6a69dd1b5aa)and I can reproduce the result on ScienceQA now. It seems that the latest code has some conflicts with LLaMA-13B-v0 on ScienceQA. By the way, can I use this version of code to run finetune code on ScienceQA?

**Comment by dingning97** [comment]
I use the latest code to reproduce the experiments on ScienceQA, But the results I got were really bad.
The Llama-2-7b achieved ~66% performance and Vicuna-7b-v1.3 achieved ~62% performance.

Here is the finetuning script I used:
####################################
PROMPT_VERSION=v1
MODEL_VERSION="vicuna-7b-v1.3"

deepspeed llava/train/train.py --deepspeed ./scripts/zero2.json \
    --lora_enable True \
    --model_name_or_path ./checkpoint/$MODEL_VERSION \
    --pretrain_mm_mlp_adapter ./checkpoint//llava-$MODEL_VERSION-pretrain/mm_projector.bin \
    --vision_tower ./checkpoint/models--openai--clip-vit-large-patch14 \
    --mm_vision_select_layer -2 \
    --data_path $SQA_FOLDER/llava_train_QCM-LEPA.json \
    --image_folder $SQA_FOLDER/images/train \
    --output_dir ./checkpoint/llava-$MODEL_VERSION-finetune-sqa-lora \
    --num_train_epochs 12 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --fp16 True \
    --bf16 False \
    --tf32 False \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 5000 \
    --save_total_limit 3 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --model_max_length 2048 \
    --mm_use_im_start_end False \
    --mm_use_im_patch_token False \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --lazy_preprocess True \
    --report_to tensorboard \
    --version $PROMPT_VERSION
####################################

Could you plz tell me what's wrong @haotian-liu @CupidJay

**Comment by dingning97** [comment]
@CupidJay  Hi!  Are you able to reproduce the result using Llama-2 or vicuna-v1.3 on ScienceQA ?

**Comment by CupidJay** [comment]
> @CupidJay Hi! Are you able to reproduce the result using Llama-2 or vicuna-v1.3 on ScienceQA ?

Hi, I am using Llama-13b on ScienceQA. I can reproduce the **evaluation** result using the given checkpoint (90.92 in the paper and I got 90.66). However, I have not reproduced the **fine-tuning** results till now (I got only 89.93).

As I noted before, I get very low accuracy with the current codebase on ScienceQA and I can only reproduce the evaluation result with this [previous commit](https://github.com/haotian-liu/LLaVA/tree/08759473173cc0a9fdc852f06a04c6a69dd1b5aa). Moreover, being able to reproduce the evaluation results does not mean that the results of fine-tuning can be reproduced, because the training method seems to have changed a lot (e.g., with/without im_end_token). 

In short, there are many variables, such as the version of the code, the version of transformers, the version of the checkpoints, the way of token processing, and the training scripts. I am also in a mess now and I think we should wait for the authors to address this issue.

**Comment by haotian-liu** [comment]
Hi all, sorry for the confusion. I have updated the instructions and the evaluation scripts for the latest code base. I also trained a version based on Vicuna v1.3, and I get accuracy 91.1% with direct eval (without GPT-4 as the judge).

Note that this checkpoint uses the LCS-558K pretrained projector, instead of CC595K one, as we do not have that for Vicuna v1.3. I will train some more checkpoints if needed in the future when I have more GPUs available.

Please check out the updated [instructions, training scripts](https://github.com/haotian-liu/LLaVA/blob/main/docs/ScienceQA.md) and [checkpoints](https://github.com/haotian-liu/LLaVA/blob/main/docs/MODEL_ZOO.md#science-qa-checkpoints). Please let me know if there is any confusion. 

Also please note that we used "QCM-LEA" format to let the model directly output the answer at the end of the answer, without a second prompt, to make the whole process cleaner and more standard. Please regenerate the training data with the correct prompt format. Thanks.

**Comment by CupidJay** [comment]
> Hi all, sorry for the confusion. I have updated the instructions and the evaluation scripts for the latest code base. I also trained a version based on Vicuna v1.3, and I get accuracy 91.1% with direct eval (without GPT-4 as the judge, [results file](https://github.com/haotian-liu/LLaVA/blob/main/llava/eval/table/results/test_sqa_llava_lcs_558k_sqa_12e_vicuna_v1_3_13b.json)).
> 
> Note that this checkpoint uses the LCS-558K pretrained projector, instead of CC595K one, as we do not have that for Vicuna v1.3. I will train some more checkpoints if needed in the future when I have more GPUs available.
> 
> Please check out the updated [instructions, training scripts](https://github.com/haotian-liu/LLaVA/blob/main/docs/ScienceQA.md) and [checkpoints](https://github.com/haotian-liu/LLaVA/blob/main/docs/MODEL_ZOO.md#science-qa-checkpoints). Please let me know if there is any confusion.
> 
> Also please note that we used "QCM-LEA" format to let the model directly output the answer at the end of the answer, without a second prompt, to make the whole process cleaner and more standard. Please regenerate the training data with the correct prompt format. Thanks.

Thanks for your efforts and it works for me now. I will close this issue.
